File: add_files_to_xcode.rb
#!/usr/bin/env ruby
# Script to add missing Swift files to Xcode project using xcodeproj gem

begin
  require 'xcodeproj'
rescue LoadError
  puts "⚠️  xcodeproj gem not found. Installing..."
  system('gem install xcodeproj') || (puts "❌ Failed to install xcodeproj gem. Please run: gem install xcodeproj" && exit(1))
  require 'xcodeproj'
end

project_path = 'Rockout.xcodeproj'
project = Xcodeproj::Project.open(project_path)

# Find the main group (Rockout)
main_group = project.main_group
rockout_group = main_group['Rockout'] || main_group

# Files to add
files_to_add = [
  'Rockout/Utils/Logger.swift',
  'Rockout/Utils/Analytics.swift',
  'Rockout/Utils/PerformanceMetrics.swift',
  'Rockout/Services/Networking/RequestCoalescer.swift',
  'Rockout/Services/Networking/RetryPolicy.swift'
]

# Find or create Utils group
utils_group = rockout_group['Utils'] || rockout_group.new_group('Utils', 'Rockout/Utils')

# Find or create Services/Networking group
services_group = rockout_group['Services'] || rockout_group.new_group('Services', 'Rockout/Services')
networking_group = services_group['Networking'] || services_group.new_group('Networking', 'Rockout/Services/Networking')

# Get the main target
target = project.targets.find { |t| t.name == 'Rockout' }
unless target
  puts "❌ Could not find 'Rockout' target"
  exit(1)
end

# Add files
files_to_add.each do |file_path|
  filename = File.basename(file_path)
  
  # Check if file already exists in project
  existing_file = project.files.find { |f| f.path == filename || f.path == file_path }
  if existing_file
    puts "✓ #{filename} already in project"
    next
  end
  
  # Determine which group to add to
  if file_path.include?('Utils/')
    group = utils_group
  elsif file_path.include?('Services/Networking/')
    group = networking_group
  else
    group = rockout_group
  end
  
  # Add file reference
  file_ref = group.new_reference(file_path)
  
  # Add to target's compile sources
  target.add_file_references([file_ref])
  
  puts "✓ Added #{filename} to project"
end

# Save the project
project.save
puts "\n✅ Project saved successfully!"

File: add_files_to_xcode.py
#!/usr/bin/env python3
"""
Script to add missing Swift files to Xcode project
"""
import re
import uuid
import os

def generate_uuid():
    """Generate a 24-character hex string for Xcode project IDs"""
    return ''.join([format(ord(c), '02X') for c in os.urandom(12)])

def add_file_to_project(project_path, file_path, group_path):
    """Add a Swift file to the Xcode project"""
    with open(project_path, 'r') as f:
        content = f.read()
    
    # Files to add
    files_to_add = [
        'Rockout/Utils/Logger.swift',
        'Rockout/Utils/Analytics.swift',
        'Rockout/Utils/PerformanceMetrics.swift',
        'Rockout/Services/Networking/RequestCoalescer.swift',
        'Rockout/Services/Networking/RetryPolicy.swift'
    ]
    
    # Check which files are already in the project
    existing_files = set()
    for match in re.finditer(r'(\w+) \/\* ([^*]+\.swift) \*\/ = \{', content):
        existing_files.add(match.group(2))
    
    print("Files already in project:")
    for f in sorted(existing_files):
        if any(x in f for x in ['Logger', 'RequestCoalescer', 'PerformanceMetrics', 'Analytics', 'RetryPolicy']):
            print(f"  ✓ {f}")
    
    print("\nFiles to add:")
    files_to_add_filtered = []
    for f in files_to_add:
        filename = os.path.basename(f)
        if filename not in existing_files:
            print(f"  + {f}")
            files_to_add_filtered.append(f)
        else:
            print(f"  ✓ {f} (already exists)")
    
    if not files_to_add_filtered:
        print("\nAll files are already in the project!")
        return
    
    print(f"\n⚠️  Note: This script can detect missing files but cannot safely modify .pbxproj")
    print("Please add these files manually in Xcode:")
    for f in files_to_add_filtered:
        print(f"  - {f}")

if __name__ == '__main__':
    project_path = 'Rockout.xcodeproj/project.pbxproj'
    if os.path.exists(project_path):
        add_file_to_project(project_path, None, None)
    else:
        print(f"Project file not found: {project_path}")

